presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleClient.java

```diff
@@ -0,0 +1,232 @@
+    }
+
+    @Override
+    public PreparedStatement getPreparedStatement(Connection connection, String sql)
+            throws SQLException
+    {
+        PreparedStatement statement = connection.prepareStatement(sql);
+        statement.setFetchSize(fetchSize);
+        return statement;
+    }
+
+    @Override
+    protected String generateTemporaryTableName()
+    {
+        return "presto_tmp_" + System.nanoTime();
+    }
+
+    @Override
+    protected void renameTable(JdbcIdentity identity, String catalogName, String schemaName, String tableName, SchemaTableName newTable)
+    {
+        if (!schemaName.equalsIgnoreCase(newTable.getSchemaName())) {
+            throw new PrestoException(NOT_SUPPORTED, "Table rename across schemas is not supported in Oracle");
+        }
+
+        String sql = format(
+                "ALTER TABLE %s RENAME TO %s",
+                quoted(catalogName, schemaName, tableName),
+                quoted(newTable.getTableName()));
+
+        try (Connection connection = connectionFactory.openConnection(identity)) {
+            execute(connection, sql);
+        }
+        catch (SQLException e) {
+            throw new PrestoException(JDBC_ERROR, e);
+        }
+    }
+
+    @Override
+    public void createSchema(JdbcIdentity identity, String schemaName)
+    {
+        // ORA-02420: missing schema authorization clause
+        throw new PrestoException(NOT_SUPPORTED, "This connector does not support creating schemas");
+    }
+
+    @Override
+    public Optional<ColumnMapping> toPrestoType(ConnectorSession session, Connection connection, JdbcTypeHandle typeHandle)
+    {
+        int columnSize = typeHandle.getColumnSize();
+
+        switch (typeHandle.getJdbcType()) {
+            case Types.CLOB:
+                return Optional.of(varcharColumnMapping(createUnboundedVarcharType()));
+            case Types.SMALLINT:
+                return Optional.of(smallintColumnMapping());
+            case Types.NUMERIC:
+                int precision = columnSize == 0 ? Decimals.MAX_PRECISION : columnSize;
+                int scale = typeHandle.getDecimalDigits();
+
+                if (scale == 0) {
+                    return Optional.of(bigintColumnMapping());
+                }
+                if (scale < 0 || scale > precision) {
+                    return Optional.of(decimalColumnMapping(createDecimalType(precision, numberDefaultScale), numberRoundingMode));
+                }
+
+                return Optional.of(decimalColumnMapping(createDecimalType(precision, scale), numberRoundingMode));
+            case Types.LONGVARCHAR:
+                if (columnSize > VarcharType.MAX_LENGTH || columnSize == 0) {
+                    return Optional.of(varcharColumnMapping(createUnboundedVarcharType()));
+                }
+                return Optional.of(varcharColumnMapping(createVarcharType(columnSize)));
+            case Types.DATE:
+                return Optional.of(timestampColumnMapping(session));
```

Comment:
Can we map to DATE type instead of TIMESTAMP using typeHandle.jdbcTypeName or typeHandle.decimalDigits? As far as I know, Oracle JDBC driver sets `DATE ` and `0`.  Please let me know if my understanding is wrong.

Reply:
Sorry, please ignore my question, the mapping itself is fine. 

But, could you confirm if we really want the condition again? Oracle won't return Types.DATE for DATE type accorting to the document and exactly I got Types.TIMESTAMP.
https://docs.oracle.com/cd/B19306_01/java.102/b14188/datamap.htm

Reply:
I think I add this by one case, but I dont remember what case and cant find it.
I remove it